### ISSUE TITLE
First-class Binary Artifacts

### ISSUE THREAD

**Issue by landonf** [CocoaPods/CocoaPods#1696]
TL/DR RFE: Support for shipping a binary, the sources corresponding to the binary,  the dSYM(s) that can be used to debug the binary, and the documentation set, if any.

Every so often, the question of providing a podspec for PLCrashReporter comes up. The CocoaPods ideal seems to be including and building from source as to allow for integration-side introspection/debugging of dependencies.

When it comes to a project like PLCrashReporter, the standard CocoaPods approach to creates some hurdles for a would-be podspec author; I've detailed the issues -- along with a pie in the sky idea on how CocoaPods could tackle them -- here.

* The build process relies on a number of custom build rules and source file references:
    * mach_exc.defs/exc.defs are included by way of a SDK-relative path reference.
    * The MIG generator uses a custom target that works around bugs in Apple's MIG target in Xcode <5.0 that breaks on projects whose absolute path contains spaces.
    * The .proto files generate protobuf-parsing code automatically via an embedded protoc implementation.
* Creating a correct build relies on a number of project-level custom configuration flags and #defines.These may vary across releases, and aren't documented or considered externally supported. if the integrator gets these wrong, it's entirely possible to create a broken or misconfigured build accidentally, and it won't be readily apparent.
* The iOS binaries are built as MH_OBJECT files and symbol stripping is performed to hide protobuf-c's library symbols from the result and avoid symbol conflicts with other libraries/applications using protobuf-c.
* A *lot* of source code is included in the project, but only a subset may apply to a given target architecture+platform; including the correct sources can be fiddly.

Most of these are problems that can be worked around, but there's a lot of room for error, and the author of the podspec will have to be very familiar with -- and reproduce --  a lot of state and logic that is encoded in the project.

As an alternative, podspec authors could include the binary .frameworks we ship directly, but then lose the introspection ability.

What I'd loosely propose is a first-class approach to shipping binary podspecs, whereby:

* A binary .framework or static library is provided.
* The exact source code *and* dSYM(s) corresponding to the binary are also included to allow source-level debugging, for open-source projects. This would ideally also included generated source code, if any.
* The documentation set(s), if any, are included.
* A standard file format (eg, a ZIP + metadata?) is defined for library authors that wish to package their build artifacts for easy inclusion with CocoaPods; I could add this to our automated release process. 

LLDB supports remapping of dSYM source paths via embedded UUID property lists: http://lldb.llvm.org/symbols.html

CocoaPods could inject the source+dSYM+binary into the project, insert or modify a UUID property list in the dSYM, and actually link against the binary artifact, with no loss of debuggability or build fidelity.

**Comment by segiddins** [comment]
I'm unclear on what actually needs to be implemented here?

**Comment by segiddins** [comment]
Closing as I can't see anything concrete to address here other than what's in https://github.com/CocoaPods/CocoaPods/issues/1698.
